src/mjolnir/landmark_database_builder.cc

```diff
@@ -1,103 +1,114 @@
+      // make the table
+      const char* table =
+          "SELECT InitSpatialMetaData(1); CREATE TABLE IF NOT EXISTS landmarks (name TEXT, type TEXT)";
+      ret = sqlite3_exec(db, table, NULL, NULL, &err_msg);
+      if (ret != SQLITE_OK) {
+        sqlite3_free(err_msg);
+        throw std::runtime_error("Sqlite table creation error: " + std::string(err_msg));
+      }
+
+      // add geom column
+      const char* geom = "SELECT AddGeometryColumn('landmarks', 'geom', 4326, 'POINT', 2)";
+      ret = sqlite3_exec(db, geom, NULL, NULL, &err_msg);
+      if (ret != SQLITE_OK) {
+        sqlite3_free(err_msg);
+        throw std::runtime_error("Sqlite geom column creation error: " + std::string(err_msg));
+      }
+
+      // make the index
+      const char* index = "SELECT CreateSpatialIndex('landmarks', 'geom')";
+      ret = sqlite3_exec(db, index, NULL, NULL, &err_msg);
+      if (ret != SQLITE_OK) {
+        sqlite3_free(err_msg);
+        throw std::runtime_error("Sqlite spatial index creation error: " + std::string(err_msg));
+      }
+
+      // prep the insert statement
+      const char* insert =
+          "INSERT INTO landmarks (name, type, geom) VALUES (?, ?, MakePoint(?, ?, 4326))";
+      ret = sqlite3_prepare_v2(db, insert, strlen(insert), &insert_stmt, NULL);
+      if (ret != SQLITE_OK)
+        throw std::runtime_error("Sqlite prepared insert statement error: " +
+                                 std::string(sqlite3_errmsg(db)));
+    }
 
-  ret = sqlite3_exec(db, sql.c_str(), NULL, NULL, &err_msg);
-  if (ret != SQLITE_OK) {
-    LOG_ERROR("Error: " + std::string(err_msg));
-    sqlite3_free(err_msg);
-    sqlite3_close(db);
-    return false;
+    // prep the select statement
+    const char* select =
+        "SELECT name, type, X(geom), Y(geom) FROM landmarks WHERE ST_Covers(BuildMbr(?, ?, ?, ?, 4326), geom)";
+    ret = sqlite3_prepare_v2(db, select, strlen(select), &bounding_box_stmt, NULL);
+    if (ret != SQLITE_OK) {
+      throw std::runtime_error("Sqlite prepared select statement error: " +
+                               std::string(sqlite3_errmsg(db)));
+    }
   }
+  ~db_pimpl() {
+    char* err_msg = nullptr;
+    if (vacuum_analyze && sqlite3_exec(db, "VACUUM", NULL, NULL, &err_msg) != SQLITE_OK) {
+      sqlite3_free(err_msg);
+      LOG_ERROR("Sqlite vacuum error: " + std::string(err_msg));
+    }
 
-  LOG_INFO("added geometry column");
-  return true;
-}
+    if (vacuum_analyze && sqlite3_exec(db, "ANALYZE", NULL, NULL, &err_msg) != SQLITE_OK) {
+      sqlite3_free(err_msg);
+      LOG_ERROR("Sqlite analyze error: " + std::string(err_msg));
+    }
 
-bool LandmarkDatabase::create_spatial_index() {
-  sql = "SELECT CreateSpatialIndex('landmarks', 'geom')";
-  ret = sqlite3_exec(db, sql.c_str(), NULL, NULL, &err_msg);
-  if (ret != SQLITE_OK) {
-    LOG_ERROR("Error: " + std::string(err_msg));
-    sqlite3_free(err_msg);
-    sqlite3_close(db);
-    return false;
+    sqlite3_finalize(insert_stmt);
```

Comment:
here you can see we finalize the statements and then close the connection

Reply:
sure. I am actually wondering why we should explicitly finalize statements in destructor, because I thought they are going to be cleaned automatically when the object is destroyed, like before throwing we dont have to close the connection.

Reply:
like the db connection the prepared statements are pointing to some memory that was dynamically allocated. in order for us to not "leak" memory we do need to manually deallocate that memory. when we `throw` the call immediately bubbles that exception upwards however the rules about what happens to objects when their life time ends still apply. that means if the outside scope that caused the exception was the scope that owned the `LandmarkDatabase` object lets that object go out of scope, its destructor (here) will be called and make sure to clean up the connection and the prepared statements.

if you mean closing the connection should finalize any statements, i dont think so. prepared statemnts are not tied to any connection they are a wholly separate construct. from the manual:

> The application must finalize every prepared statement in order to avoid resource leaks.

Reply:
oh I see. I mixed up some concepts earlier but now I understand everything.
thanks for the detailed explanation!